File: framework/scala/src/main/scala/soda/scala/unittest/TestWork.scala
package soda.scala.unittest

import scala.reflect.runtime.universe._

import soda.scala.unittest.task.ReflectionTask

class TestWork(methodMirror: MethodMirror) {

  private val gwork = new GenericTestWork[Any](new ReflectionTask(methodMirror))

  var compareSerial: Boolean = false

  def run(text: String): String = {
    gwork.compareSerial = compareSerial
    gwork.run(text)
  }

  def run(): Unit = {
    println(run(Utils.fromStdin()))
  }

  def setValidator(v: (_, _) => Boolean): Unit = {
    gwork.setValidator(v.asInstanceOf[(Any,Any)=>Boolean])
  }

  def arguments: List[Any] = gwork.arguments
}

object TestWork {

  def forObject(objType: Type, methodName: String): TestWork = {
    val rMirror = runtimeMirror(Thread.currentThread().getContextClassLoader)
    val instMirror = rMirror.reflect(
      rMirror.reflectModule(objType.typeSymbol.asClass.companion.asModule).instance
    )
    val mm = instMirror.reflectMethod(
      objType.companion.decl(TermName(methodName)).asMethod
    )
    new TestWork(mm)
  }

  def forInstance(inst: Any, methodName: String): TestWork = {
    val rMirror = runtimeMirror(Thread.currentThread.getContextClassLoader)
    val instMirror = rMirror.reflect(inst)
    val mm = instMirror.reflectMethod(
      instMirror.symbol.typeSignature.decl(TermName(methodName)).asMethod
    )
    new TestWork(mm)
  }

  def forStruct(classType: Type): TestWork = {
    val tester = new StructTester(classType)
    val methodName = "test"
    forInstance(tester, methodName)
  }
}

File: framework/java/src/main/java/soda/unittest/TestWork.java
package soda.unittest;

import java.util.function.BiPredicate;

import soda.unittest.task.ReflectionTask;

public class TestWork {

	private final GenericTestWork<Object> gwork;

	public TestWork(Object su, String methodName) {
		gwork = new GenericTestWork<>(new ReflectionTask(su, methodName));
	}

	public void setCompareSerial(boolean b) {
		gwork.setCompareSerial(b);
	}
	
	public void setValidator(BiPredicate<?,?> v) {
		gwork.setValidator(Utils.cast(v));
	}

	public Object[] getArguments() {
		return gwork.getArguments();
	}

	public void run() {
		Utils.wrapEx(() -> System.out.println(run(Utils.fromStdin())));
	}

	public String run(String text) {
		return gwork.run(text);
	}

	public static TestWork forStruct(Class<?> structClass) {
		return StructTester.createTestWork(structClass);
	}

	public static TestWork forObject(Object obj, String method) {
		return new TestWork(obj, method);
	}
	
}
